doc/neps/nep-0041-improved-dtype-support.rst

```diff
@@ -0,0 +1,706 @@
+
+   * Currently user dtypes are defined as instances of ``np.dtype``.
+     The creation works by the user providing a prototype instance.
+     NumPy will need to modify at least the type during registration.
+     This has no effect for either ``rational`` or ``quaternion`` and mutation
+     of the structure seems unlikely after registration.
+
+Since there is a fairly large API surface concerning datatypes, further changes
+or the limitation of a certain function to currently existing datatypes is
+likely to occur.
+For example functions which use the type number as input
+should be replaced with functions taking DType classes instead.
+Although public, large parts of this C-API seems very rarely and possibly
+never used by downstream projects.
+
+
+
+
+Detailed Description
+--------------------
+
+This section details some of the design decisions above and gives
+more details.
+Since datatype changes touch a large part of code and behaviours, NEP 40
+reviews some of the concepts, issues, and the current implementation.
+
+
+Datatypes as Python Classes (1)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The current NumPy datatypes are not full scale python classes.
+They are instead (prototype) instances of a single ``np.dtype`` class.
+Changing this means that any special handling, e.g. for ``datetime``
+can be moved to the Datetime DType class instead, away from monolithic general
+code (e.g. current ``PyArray_AdjustFlexibleDType``).
+
+The main API side effect of this is that special methods move
+from the dtype instances to class methods on the class.
+This is the typical design pattern used in Python.
+Adding a new, pythonic point to store these methods and information, will
+allow us to further refine the API and allow it to grow in the future.
+The current API cannot be extended due to how it is exposed publically.
+This means for example that the methods currently stored in ``ArrFunctions``
+on each datatype (see NEP 40) will be defined differently in the future and
+deprecated in the long run.
+
+The most prominent visible side effect of this will be that
+``type(np.dtype(np.float64))`` will not be ``np.dtype`` anymore.
+Instead it will be a subclass of ``np.dtype`` meaning that
+``isinstance(np.dtype(np.float64), np.dtype)`` will remain true.
+This will also add the ability to use ``isinstance(dtype, np.dtype[float64])``
+thus removing the need to use ``dtype.kind``, ``dtype.char``, or ``dtype.type``
+to do this check.
+
+If DTypes were full scale Python classes, the question of subclassing arises.
+Inheritance, however, appears problematic and a complexity best avoided
+(at least initially) for container datatypes.
+Further, subclasses may be more interesting for the interoperatbility for
+example with GPU backends (CuPy) storing additional methods related to the
+GPU rather than as a mechanism to define new datatypes.
+Since a class hierarchy and subclass order provides value, phase I will allow
```

Comment:
"subclass order" -> "method resolution order"?

Reply:
Removed mro does not make sense here